packages/engine-cli/src/node-config-manager.ts

```diff
@@ -0,0 +1,140 @@
+import crypto from 'node:crypto';
+import esbuild from 'esbuild';
+import { runInContext, createContext } from 'node:vm';
+
+type BuildStats = {
+    error: Error | undefined;
+    currentValue: Promise<unknown[]> | unknown[];
+    dispose(): Promise<void>;
+    build?: ReturnType<typeof deferred<unknown[]>>;
+};
+
+export class NodeConfigManager {
+    constructor(private config: esbuild.BuildOptions) {}
+    runningBuilds = new Map<string, BuildStats>();
+    hashConfig(entryPoints: string[]) {
+        return hash(entryPoints.join(','));
+    }
+    async loadConfigs(entryPoints: string[]) {
+        const key = this.hashConfig(entryPoints);
+        const currentBuild = this.runningBuilds.get(key);
+        if (!currentBuild) {
+            const buildStats: BuildStats = {
+                error: undefined,
+                currentValue: Promise.reject(new Error('No build')),
+                build: undefined,
+                dispose() {
+                    return ctx.dispose();
+                },
+            };
+            this.runningBuilds.set(key, buildStats);
+
+            const ctx = await this.createBuildTask(entryPoints, {
+                onStart() {
+                    const newBuild = deferred<unknown[]>();
+                    buildStats.error = undefined;
+                    buildStats.currentValue = newBuild.promise;
+                    buildStats.build = newBuild;
+                },
+                onEnd(files) {
+                    buildStats.error = undefined;
```

Comment:
is there a case that build has an error and we still get to build end?

Reply:
nope.